src/platform/datapath_raw_socket.c

```diff
@@ -79,20 +76,91 @@ CxPlatTryAddSocket(
     _In_ CXPLAT_SOCKET* Socket
     )
 {
-    BOOLEAN Success = TRUE;
+    int Result;
+    BOOLEAN Success = FALSE;
     CXPLAT_HASHTABLE_LOOKUP_CONTEXT Context;
     CXPLAT_HASHTABLE_ENTRY* Entry;
 
+    //
+    // Get (and reserve) a transport layer port from tcpip.sys by binding an auxiliary socket.
+    //
+
+    Socket->AuxSocket =
+        WSASocketW(
+            AF_INET6,
+            SOCK_DGRAM,
+            IPPROTO_UDP,
+            NULL,
+            0,
+            WSA_FLAG_OVERLAPPED);
+    if (Socket->AuxSocket == INVALID_SOCKET) {
+        int WsaError = WSAGetLastError();
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            Socket,
+            WsaError,
+            "WSASocketW");
+        goto Error;
+    }
+
+    int Option = FALSE;
+    Result =
+        setsockopt(
+            Socket->AuxSocket,
+            IPPROTO_IPV6,
+            IPV6_V6ONLY,
+            (char*)&Option,
+            sizeof(Option));
+    if (Result == SOCKET_ERROR) {
+        int WsaError = WSAGetLastError();
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            Socket,
+            WsaError,
+            "Set IPV6_V6ONLY");
+        goto Error;
+    }
+
     CxPlatRwLockAcquireExclusive(&Pool->Lock);
 
-    if (Socket->LocalAddress.Ipv4.sin_port == 0) {
-        Socket->LocalAddress.Ipv4.sin_port = CxPlatByteSwapUint16(Pool->NextLocalPort);
-        Pool->NextLocalPort++;
-        if (Pool->NextLocalPort == 0) {
-            Pool->NextLocalPort = HARDCODED_SOCK_POOL_START_PORT;
-        }
+    Result =
+        bind(
```

Comment:
When `Socket-Connected` is `TRUE` we need to call `connect` as well.

Reply:
Why?

Reply:
To also get the right local IP address

Reply:
We need to discuss this.

Reply:
Discussed- connect sounds good!